File: frontend/src/types/sla.ts
export interface SlaPolicy {
  id: string;
  workspaceId: string;
  sectorId?: string;
  channelAccountId?: string;
  priority?: string;
  name: string;
  firstResponseMinutes: number;
  nextResponseMinutes?: number;
  resolutionMinutes: number;
  reopenResponseMinutes?: number;
  businessHoursOnly: boolean;
  pauseWhenWaitingCustomer: boolean;
  atRiskThresholdPct: number;
  steps: Record<string, unknown>;
  active: boolean;
  createdAt: string;
}

export interface AgentCapacity {
  id: string;
  workspaceId: string;
  userId: string;
  sectorId?: string;
  maxConversations: number;
  maxWeight: number;
  priorityWeights: Record<string, number>;
}

export type AgentPresence = "online" | "away" | "busy" | "in_call" | "on_break" | "offline" | "invisible";

export interface AgentStatus {
  id: string;
  workspaceId: string;
  userId: string;
  status: AgentPresence;
  reasonId?: string | null;
  note?: string | null;
  sinceAt: string;
  updatedAt: string;
  userName?: string | null;
  userEmail?: string | null;
  reasonLabel?: string | null;
}

export interface AgentPauseReason {
  id: string;
  workspaceId: string;
  label: string;
  active: boolean;
  position: number;
}

export interface BusinessHours {
  id: string;
  workspaceId: string;
  sectorId?: string | null;
  weekday: number;
  startMinute: number;
  endMinute: number;
  timezone: string;
  active: boolean;
}

export interface RoutingRule {
  id: string;
  workspaceId: string;
  sectorId?: string | null;
  strategy: "round_robin" | "least_busy" | "sticky_agent" | "manual";
  tiebreaker: string;
  stickyHours: number;
  autoReassignMinutes: number;
  reopenWindowHours: number;
}

export interface SupervisorSectorMetric {
  sectorId?: string | null;
  sectorName: string;
  queued: number;
  active: number;
  atRisk: number;
  violated: number;
}

export interface SupervisorAgentMetric {
  userId: string;
  name: string;
  email: string;
  role: string;
  sectorId?: string | null;
  status: AgentPresence;
  reason?: string | null;
  sinceAt?: string | null;
  maxConversations: number;
  maxWeight: number;
  assignedOpen: number;
  weightedLoad: number;
  availableSlots: number;
  atRisk: number;
  violated: number;
}

export interface SupervisorAlert {
  conversationId: string;
  contactName?: string | null;
  assigneeId?: string | null;
  sectorId?: string | null;
  slaStatus: "ok" | "at_risk" | "violated";
  priority: string;
  lastMessageAt?: string | null;
}

export interface SupervisorOverview {
  totals: {
    queued: number;
    active: number;
    atRisk: number;
    violated: number;
    agentsOnline: number;
    capacityUsed: number;
    capacityTotal: number;
  };
  sectors: SupervisorSectorMetric[];
  agents: SupervisorAgentMetric[];
  alerts: SupervisorAlert[];
}

// ── Stage 9 extras types ────────────────────────────────────────────────────

export interface AutoResolveRule {
  id: string;
  workspaceId: string;
  sectorId?: string | null;
  inactivityHours: number;
  statusFrom: string[];
  statusTo: string;
  active: boolean;
}

export interface BusinessHoliday {
  id: string;
  workspaceId: string;
  sectorId?: string | null;
  holidayDate: string;
  label: string;
  treatAs: "closed" | "half_day" | "custom";
  customStartMinute?: number | null;
  customEndMinute?: number | null;
}

export type EscalationAction = "notify" | "reassign" | "webhook";

export interface SlaEscalationRule {
  id?: string;
  workspaceId?: string;
  policyId?: string;
  thresholdPct: number;
  action: EscalationAction;
  targetRole?: string | null;
  targetUserId?: string | null;
  webhookUrl?: string | null;
  position: number;
  active: boolean;
}

export interface IdleRule {
  id: string;
  workspaceId: string;
  idleMinutes: number;
  offlineMinutes: number;
  active: boolean;
}

export type NotificationKind = "email" | "slack_webhook" | "inapp" | "webhook";

export interface NotificationChannel {
  id: string;
  workspaceId: string;
  name: string;
  kind: NotificationKind;
  config: Record<string, unknown>;
  events: string[];
  active: boolean;
  createdAt: string;
}

export interface NotificationDelivery {
  id: string;
  workspaceId: string;
  channelId?: string | null;
  userId?: string | null;
  eventType: string;
  title: string;
  body?: string | null;
  status: string;
  readAt?: string | null;
  createdAt: string;
}

export interface ExternalWebhookSubscription {
  id: string;
  workspaceId: string;
  name: string;
  url: string;
  events: string[];
  active: boolean;
  createdAt: string;
}

export interface ExternalWebhookDelivery {
  id: string;
  subscriptionId: string;
  eventType: string;
  status: string;
  attempts: number;
  lastError?: string | null;
  responseStatus?: number | null;
  deliveredAt?: string | null;
  createdAt: string;
}

export interface ConversationLock {
  id: string;
  conversationId: string;
  holderUserId: string;
  acquiredAt: string;
  expiresAt: string;
  holderName?: string | null;
}

export interface CsatSummary {
  items: Array<{
    id: string;
    conversationId: string;
    assigneeId?: string | null;
    score?: number | null;
    feedback?: string | null;
    sentAt: string;
    respondedAt?: string | null;
  }>;
  averageScore?: number | null;
}

export interface HeatmapCell {
  userId: string;
  userName: string;
  hour: number;
  minutesAssigned: number;
  conversations: number;
}

export interface HeatmapData {
  days: number;
  cells: HeatmapCell[];
}

File: backend/app/schemas/sla.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


AgentPresence = Literal["online", "away", "busy", "in_call", "on_break", "offline", "invisible"]
RoutingStrategy = Literal["round_robin", "least_busy", "sticky_agent", "manual"]
EscalationAction = Literal["notify", "reassign", "webhook"]


class SlaPolicyCreate(BaseModel):
    name: str
    sector_id: uuid.UUID | None = None
    channel_account_id: uuid.UUID | None = None
    priority: str | None = None
    first_response_minutes: int
    next_response_minutes: int | None = None
    resolution_minutes: int
    reopen_response_minutes: int | None = None
    business_hours_only: bool = True
    pause_when_waiting_customer: bool = True
    at_risk_threshold_pct: int = 80
    steps: dict = {}
    active: bool = True


class SlaPolicyUpdate(BaseModel):
    name: str | None = None
    sector_id: uuid.UUID | None = None
    channel_account_id: uuid.UUID | None = None
    priority: str | None = None
    first_response_minutes: int | None = None
    next_response_minutes: int | None = None
    resolution_minutes: int | None = None
    reopen_response_minutes: int | None = None
    business_hours_only: bool | None = None
    pause_when_waiting_customer: bool | None = None
    at_risk_threshold_pct: int | None = None
    steps: dict | None = None
    active: bool | None = None


class SlaPolicyOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    sector_id: uuid.UUID | None
    channel_account_id: uuid.UUID | None
    priority: str | None
    name: str
    first_response_minutes: int
    next_response_minutes: int | None
    resolution_minutes: int
    reopen_response_minutes: int | None
    business_hours_only: bool
    pause_when_waiting_customer: bool
    at_risk_threshold_pct: int
    steps: dict
    active: bool
    created_at: datetime


class AgentCapacitySet(BaseModel):
    user_id: uuid.UUID
    max_conversations: int
    sector_id: uuid.UUID | None = None
    max_weight: float | None = None
    priority_weights: dict | None = None


class AgentCapacityOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    user_id: uuid.UUID
    sector_id: uuid.UUID | None
    max_conversations: int
    max_weight: float
    priority_weights: dict


class AgentPauseReasonCreate(BaseModel):
    label: str
    active: bool = True
    position: int = 0


class AgentPauseReasonOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    label: str
    active: bool
    position: int


class AgentStatusSet(BaseModel):
    status: AgentPresence
    reason_id: uuid.UUID | None = None
    note: str | None = None


class AgentStatusOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    user_id: uuid.UUID
    status: str
    reason_id: uuid.UUID | None
    note: str | None
    since_at: datetime
    updated_at: datetime
    user_name: str | None = None
    user_email: str | None = None
    reason_label: str | None = None


class BusinessHoursIn(BaseModel):
    sector_id: uuid.UUID | None = None
    weekday: int
    start_minute: int
    end_minute: int
    timezone: str = "America/Sao_Paulo"
    active: bool = True


class BusinessHoursOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    sector_id: uuid.UUID | None
    weekday: int
    start_minute: int
    end_minute: int
    timezone: str
    active: bool


class RoutingRuleIn(BaseModel):
    sector_id: uuid.UUID | None = None
    strategy: RoutingStrategy = "least_busy"
    tiebreaker: str = "oldest_idle"
    sticky_hours: int = 24
    auto_reassign_minutes: int = 10
    reopen_window_hours: int = 24


class RoutingRuleOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    sector_id: uuid.UUID | None
    strategy: str
    tiebreaker: str
    sticky_hours: int
    auto_reassign_minutes: int
    reopen_window_hours: int


class SlaEscalationRuleIn(BaseModel):
    threshold_pct: int
    action: EscalationAction
    target_role: str | None = None
    target_user_id: uuid.UUID | None = None
    webhook_url: str | None = None
    position: int = 0
    active: bool = True


class SlaEscalationRuleOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    policy_id: uuid.UUID
    threshold_pct: int
    action: str
    target_role: str | None
    target_user_id: uuid.UUID | None
    webhook_url: str | None
    position: int
    active: bool


class AssignmentRequest(BaseModel):
    conversation_id: uuid.UUID
    sector_id: uuid.UUID | None = None


class AssignmentResult(BaseModel):
    conversation_id: uuid.UUID
    assignee_id: uuid.UUID | None
    assigned: bool
    reason: str | None = None


class SlaEventOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    workspace_id: uuid.UUID
    conversation_id: uuid.UUID
    policy_id: uuid.UUID
    type: str
    status: str
    deadline_at: datetime
    achieved_at: datetime | None
    violated_at: datetime | None
    escalated_to: uuid.UUID | None
    metadata_: dict


class SupervisorSectorMetric(BaseModel):
    sector_id: uuid.UUID | None
    sector_name: str
    queued: int
    active: int
    at_risk: int
    violated: int


class SupervisorAgentMetric(BaseModel):
    user_id: uuid.UUID
    name: str
    email: str
    role: str
    sector_id: uuid.UUID | None = None
    status: str
    reason: str | None = None
    since_at: datetime | None = None
    max_conversations: int
    max_weight: float
    assigned_open: int
    weighted_load: float
    available_slots: float
    at_risk: int
    violated: int


class SupervisorAlert(BaseModel):
    conversation_id: uuid.UUID
    contact_name: str | None
    assignee_id: uuid.UUID | None
    sector_id: uuid.UUID | None
    sla_status: str
    priority: str
    last_message_at: datetime | None


class SupervisorOverview(BaseModel):
    totals: dict
    sectors: list[SupervisorSectorMetric]
    agents: list[SupervisorAgentMetric]
    alerts: list[SupervisorAlert]
